File: backend/app/excel_parser.py
"""Parser dinámico para archivos Excel de rutas Pan Pa Ya.

Estructura del Excel:
- Fila 1: fecha de entrega
- Bloques bidireccionales en columnas A-D (izq) y F-I (der), separadas por col E
- Cada bloque: encabezado de zona → filas de clientes → NOMBRE / PLACA / AUX.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import date, datetime, time
from pathlib import Path
from typing import BinaryIO

import openpyxl

# Columnas 0-indexed
COL_IZQ = (0, 1, 2, 3)  # código, nombre, dirección, notas
COL_DER = (5, 6, 7, 8)
COL_SEP = 4

DRIVER_LABELS = {"NOMBRE", "PLACA", "AUX.", "AUX"}
CODIGO_PATTERN = re.compile(r"^\d{4,7}$")


@dataclass
class PedidoParseado:
    codigo_cliente: str
    nombre_cliente: str
    direccion: str
    notas: str | None = None
    orden: int | None = None


@dataclass
class RutaParseada:
    fecha: date
    nombre_zona: str
    lado: str  # 'izquierda' | 'derecha'
    conductor: str
    placa: str
    auxiliar: str | None = None
    hora_salida: time | None = None
    pedidos: list[PedidoParseado] = field(default_factory=list)


def _cell_str(value) -> str:
    if value is None:
        return ""
    if isinstance(value, datetime):
        return value.strftime("%Y-%m-%d")
    if isinstance(value, time):
        return value.strftime("%H:%M:%S")
    return str(value).strip()


def _parse_fecha(value) -> date:
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    text = _cell_str(value)
    for fmt in ("%Y-%m-%d", "%d.%m.%Y", "%d/%m/%Y"):
        try:
            return datetime.strptime(text[:10], fmt).date()
        except ValueError:
            continue
    raise ValueError(f"No se pudo interpretar la fecha: {value!r}")


def _parse_hora(value) -> time | None:
    if value is None:
        return None
    if isinstance(value, time):
        return value
    if isinstance(value, datetime):
        return value.time()
    text = _cell_str(value)
    if not text:
        return None
    for fmt in ("%H:%M:%S", "%H:%M"):
        try:
            return datetime.strptime(text, fmt).time()
        except ValueError:
            continue
    return None


def _parse_orden(value) -> int | None:
    text = _cell_str(value)
    if not text:
        return None
    try:
        return int(float(text))
    except ValueError:
        return None


def _is_codigo_cliente(text: str) -> bool:
    return bool(CODIGO_PATTERN.match(text))


def _is_driver_row(label: str) -> bool:
    return label.upper() in DRIVER_LABELS or label.upper().startswith("AUX")


def _is_zone_header(row: list[str], col: int) -> bool:
    """Detecta encabezado de zona en una columna."""
    label = row[col] if col < len(row) else ""
    if not label:
        return False
    if _is_codigo_cliente(label):
        return False
    if _is_driver_row(label):
        return False
    if label.upper() in ("DIRECCION", "DIRECCIÓN", "ORDEN", "MULTIPLAZA"):
        return False
    return True


def _detect_zone(row: list[str], side: str) -> str | None:
    """Detecta nombre de zona en layout estándar (col 0/5) o MULTIAMBIENTE (col 1/6)."""
    if side == "izquierda":
        cols = (0, 1)
        opp_base = 5
    else:
        cols = (5, 6)
        opp_base = 0

    for col in cols:
        if not _is_zone_header(row, col):
            continue
        # Título desplazado (MULTIAMBIENTE): clientes siguen en col base
        if col in (1, 6) and row[cols[0]]:
            continue
        # Si el otro lado tiene código de cliente, es etiqueta interna, no nueva zona
        if _is_codigo_cliente(row[opp_base]):
            continue
        return row[col]
    return None


def _row_starts_new_block(row: list[str]) -> bool:
    return _detect_zone(row, "izquierda") is not None or _detect_zone(row, "derecha") is not None


def _read_row(ws, row_idx: int, max_col: int = 9) -> list[str]:
    return [_cell_str(ws.cell(row=row_idx, column=c + 1).value) for c in range(max_col)]


def _extract_cliente(row: list[str], cols: tuple[int, ...]) -> PedidoParseado | None:
    codigo = row[cols[0]] if cols[0] < len(row) else ""
    if not _is_codigo_cliente(codigo):
        return None
    nombre = row[cols[1]] if cols[1] < len(row) else ""
    direccion = row[cols[2]] if cols[2] < len(row) else ""
    notas = row[cols[3]] if cols[3] < len(row) else ""
    orden_val = _parse_orden(notas)
    notas_final = notas if notas and orden_val is None else None
    return PedidoParseado(
        codigo_cliente=codigo,
        nombre_cliente=nombre,
        direccion=direccion,
        notas=notas_final if notas_final else None,
        orden=orden_val,
    )


def _parse_block(
    rows: list[list[str]],
    col_start: int,
    lado: str,
    fecha: date,
    nombre_zona: str,
) -> RutaParseada | None:
    pedidos: list[PedidoParseado] = []
    conductor = ""
    placa = ""
    auxiliar: str | None = None
    hora_salida: time | None = None

    for row in rows:
        label = row[col_start] if col_start < len(row) else ""

        if _is_driver_row(label):
            valor = row[col_start + 1] if col_start + 1 < len(row) else ""
            hora_celda = row[col_start + 2] if col_start + 2 < len(row) else ""
            upper = label.upper()
            if upper == "NOMBRE":
                conductor = valor
                h = _parse_hora(hora_celda)
                if h:
                    hora_salida = h
            elif upper == "PLACA":
                placa = valor
                h = _parse_hora(hora_celda)
                if h and not hora_salida:
                    hora_salida = h
            elif upper.startswith("AUX"):
                auxiliar = valor or None
                h = _parse_hora(hora_celda)
                if h and not hora_salida:
                    hora_salida = h
            continue

        cliente = _extract_cliente(row, (col_start, col_start + 1, col_start + 2, col_start + 3))
        if cliente:
            pedidos.append(cliente)

    if not conductor and not placa and not pedidos:
        return None

    return RutaParseada(
        fecha=fecha,
        nombre_zona=nombre_zona,
        lado=lado,
        conductor=conductor or "SIN ASIGNAR",
        placa=placa or "SIN PLACA",
        auxiliar=auxiliar,
        hora_salida=hora_salida,
        pedidos=pedidos,
    )


def parse_excel(source: str | Path | BinaryIO) -> list[RutaParseada]:
    """Parsea un archivo Excel y devuelve todas las rutas detectadas."""
    wb = openpyxl.load_workbook(source, data_only=True)
    ws = wb.active

    fecha = _parse_fecha(ws.cell(row=1, column=1).value)
    rutas: list[RutaParseada] = []
    max_row = ws.max_row

    row_idx = 2
    while row_idx <= max_row:
        row = _read_row(ws, row_idx)

        zona_izq = _detect_zone(row, "izquierda")
        zona_der = _detect_zone(row, "derecha")

        if not zona_izq and not zona_der:
            row_idx += 1
            continue

        block_rows: list[list[str]] = []
        row_idx += 1
        while row_idx <= max_row:
            next_row = _read_row(ws, row_idx)
            if _row_starts_new_block(next_row):
                break
            if all(not c for c in next_row) and row_idx + 1 <= max_row:
                peek = _read_row(ws, row_idx + 1)
                if _row_starts_new_block(peek):
                    break
            block_rows.append(next_row)
            row_idx += 1

        if zona_izq:
            ruta = _parse_block(block_rows, COL_IZQ[0], "izquierda", fecha, zona_izq)
            if ruta:
                rutas.append(ruta)

        if zona_der:
            ruta = _parse_block(block_rows, COL_DER[0], "derecha", fecha, zona_der)
            if ruta:
                rutas.append(ruta)

    return rutas

File: web/src/lib/excelParser.ts
/**
 * Parser dinámico Excel Pan Pa Ya — portado desde backend Python.
 * Bloques bidireccionales: columnas 0-3 (izq) y 5-8 (der).
 */

import * as XLSX from "xlsx";

const DRIVER_LABELS = new Set(["NOMBRE", "PLACA", "AUX.", "AUX"]);
const CODIGO_PATTERN = /^\d{4,7}$/;

export type PedidoParseado = {
  codigo_cliente: string;
  nombre_cliente: string;
  direccion: string;
  notas: string | null;
  orden: number | null;
};

export type RutaParseada = {
  fecha: string;
  nombre_zona: string;
  lado: "izquierda" | "derecha";
  conductor: string;
  placa: string;
  auxiliar: string | null;
  hora_salida: string | null;
  pedidos: PedidoParseado[];
};

function cellStr(value: unknown): string {
  if (value == null) return "";
  if (value instanceof Date) return value.toISOString().slice(0, 10);
  return String(value).trim();
}

function parseFecha(value: unknown): string {
  if (value instanceof Date) return value.toISOString().slice(0, 10);
  const text = cellStr(value);
  if (/^\d{4}-\d{2}-\d{2}/.test(text)) return text.slice(0, 10);
  const parts = text.match(/^(\d{1,2})[./](\d{1,2})[./](\d{4})/);
  if (parts) {
    return `${parts[3]}-${parts[2].padStart(2, "0")}-${parts[1].padStart(2, "0")}`;
  }
  if (typeof value === "number") {
    const d = XLSX.SSF.parse_date_code(value);
    if (d) {
      return `${d.y}-${String(d.m).padStart(2, "0")}-${String(d.d).padStart(2, "0")}`;
    }
  }
  throw new Error(`No se pudo interpretar la fecha: ${text}`);
}

function parseHora(value: unknown): string | null {
  if (value == null) return null;
  if (value instanceof Date) return value.toTimeString().slice(0, 8);
  const text = cellStr(value);
  if (!text) return null;
  if (/^\d{1,2}:\d{2}/.test(text)) {
    const [h, m, s = "0"] = text.split(":");
    return `${h.padStart(2, "0")}:${m.padStart(2, "0")}:${s.padStart(2, "0")}`;
  }
  if (typeof value === "number" && value < 1) {
    const totalSec = Math.round(value * 86400);
    const h = Math.floor(totalSec / 3600);
    const m = Math.floor((totalSec % 3600) / 60);
    const s = totalSec % 60;
    return `${String(h).padStart(2, "0")}:${String(m).padStart(2, "0")}:${String(s).padStart(2, "0")}`;
  }
  return null;
}

function parseOrden(value: string): number | null {
  if (!value) return null;
  const n = parseInt(value, 10);
  return Number.isNaN(n) ? null : n;
}

function isCodigoCliente(text: string): boolean {
  return CODIGO_PATTERN.test(text);
}

function isDriverRow(label: string): boolean {
  const u = label.toUpperCase();
  return DRIVER_LABELS.has(u) || u.startsWith("AUX");
}

function isZoneHeader(row: string[], col: number): boolean {
  const label = row[col] ?? "";
  if (!label) return false;
  if (isCodigoCliente(label)) return false;
  if (isDriverRow(label)) return false;
  if (["DIRECCION", "DIRECCIÓN", "ORDEN", "MULTIPLAZA"].includes(label.toUpperCase()))
    return false;
  return true;
}

function detectZone(row: string[], side: "izquierda" | "derecha"): string | null {
  const cols = side === "izquierda" ? [0, 1] : [5, 6];
  const oppBase = side === "izquierda" ? 5 : 0;

  for (const col of cols) {
    if (!isZoneHeader(row, col)) continue;
    if ((col === 1 || col === 6) && row[cols[0]]) continue;
    if (isCodigoCliente(row[oppBase] ?? "")) continue;
    return row[col];
  }
  return null;
}

function rowStartsNewBlock(row: string[]): boolean {
  return detectZone(row, "izquierda") != null || detectZone(row, "derecha") != null;
}

function normalizeRow(fila: unknown[]): string[] {
  const row: string[] = [];
  for (let c = 0; c < 9; c++) {
    row.push(cellStr(fila[c]));
  }
  return row;
}

function extractCliente(row: string[], colStart: number): PedidoParseado | null {
  const codigo = row[colStart] ?? "";
  if (!isCodigoCliente(codigo)) return null;
  const nombre = row[colStart + 1] ?? "";
  const direccion = row[colStart + 2] ?? "";
  const notasRaw = row[colStart + 3] ?? "";
  const orden = parseOrden(notasRaw);
  return {
    codigo_cliente: codigo,
    nombre_cliente: nombre,
    direccion,
    notas: notasRaw && orden == null ? notasRaw : null,
    orden,
  };
}

function parseBlock(
  rows: string[][],
  colStart: number,
  lado: "izquierda" | "derecha",
  fecha: string,
  nombreZona: string
): RutaParseada | null {
  const pedidos: PedidoParseado[] = [];
  let conductor = "";
  let placa = "";
  let auxiliar: string | null = null;
  let horaSalida: string | null = null;

  for (const row of rows) {
    const label = row[colStart] ?? "";

    if (isDriverRow(label)) {
      const valor = row[colStart + 1] ?? "";
      const horaCelda = row[colStart + 2] ?? "";
      const upper = label.toUpperCase();
      if (upper === "NOMBRE") {
        conductor = valor;
        const h = parseHora(horaCelda);
        if (h) horaSalida = h;
      } else if (upper === "PLACA") {
        placa = valor;
        const h = parseHora(horaCelda);
        if (h && !horaSalida) horaSalida = h;
      } else if (upper.startsWith("AUX")) {
        auxiliar = valor || null;
        const h = parseHora(horaCelda);
        if (h && !horaSalida) horaSalida = h;
      }
      continue;
    }

    const cliente = extractCliente(row, colStart);
    if (cliente) pedidos.push(cliente);
  }

  if (!conductor && !placa && pedidos.length === 0) return null;

  return {
    fecha,
    nombre_zona: nombreZona,
    lado,
    conductor: conductor || "SIN ASIGNAR",
    placa: placa || "SIN PLACA",
    auxiliar,
    hora_salida: horaSalida,
    pedidos,
  };
}

export function parseExcelBuffer(buffer: Buffer): RutaParseada[] {
  const workbook = XLSX.read(buffer, { type: "buffer", cellDates: true });
  const sheetName = workbook.SheetNames[0];
  const worksheet = workbook.Sheets[sheetName];
  const filasRaw = XLSX.utils.sheet_to_json<unknown[]>(worksheet, {
    header: 1,
    defval: "",
  });

  if (!filasRaw.length) throw new Error("El archivo Excel está vacío");

  const fecha = parseFecha(filasRaw[0]?.[0]);
  const rutas: RutaParseada[] = [];

  let rowIdx = 1;
  while (rowIdx < filasRaw.length) {
    const row = normalizeRow(filasRaw[rowIdx] as unknown[]);
    const zonaIzq = detectZone(row, "izquierda");
    const zonaDer = detectZone(row, "derecha");

    if (!zonaIzq && !zonaDer) {
      rowIdx++;
      continue;
    }

    const blockRows: string[][] = [];
    rowIdx++;
    while (rowIdx < filasRaw.length) {
      const nextRow = normalizeRow(filasRaw[rowIdx] as unknown[]);
      if (rowStartsNewBlock(nextRow)) break;
      if (nextRow.every((c) => !c) && rowIdx + 1 < filasRaw.length) {
        const peek = normalizeRow(filasRaw[rowIdx + 1] as unknown[]);
        if (rowStartsNewBlock(peek)) break;
      }
      blockRows.push(nextRow);
      rowIdx++;
    }

    if (zonaIzq) {
      const ruta = parseBlock(blockRows, 0, "izquierda", fecha, zonaIzq);
      if (ruta) rutas.push(ruta);
    }
    if (zonaDer) {
      const ruta = parseBlock(blockRows, 5, "derecha", fecha, zonaDer);
      if (ruta) rutas.push(ruta);
    }
  }

  return rutas;
}
